enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,901 @@
+
+#### Usage of the json representation of packages by olm components
+
+The sqllite database that is currently created inside the index is used to serve contents of the index over a gRPC api.
+
+```go
+rpc ListPackages(ListPackageRequest) returns (stream PackageName) {}
+rpc GetPackage(GetPackageRequest) returns (Package) {}
+rpc GetBundle(GetBundleRequest) returns (Bundle) {}
+rpc GetBundleForChannel(GetBundleInChannelRequest) returns (Bundle) {}
+rpc GetChannelEntriesThatReplace(GetAllReplacementsRequest) returns (stream ChannelEntry) {}
+rpc GetBundleThatReplaces(GetReplacementRequest) returns (Bundle) {}
+rpc GetChannelEntriesThatProvide(GetAllProvidersRequest) returns (stream ChannelEntry) {}
+rpc GetLatestChannelEntriesThatProvide(GetLatestProvidersRequest) returns (stream ChannelEntry) {}
+rpc GetDefaultBundleThatProvides(GetDefaultProviderRequest) returns (Bundle) {}
+rpc ListBundles(ListBundlesRequest) returns (stream Bundle) {}
+```
+
+Once the declarative package configs are available inside an index, these configs will be used to serve the content for these api endpoints instead of the sqllite database. A new sub-command `serve` will be introduced under `opm index` that will take a aggregated config file as input, and use that file to serve the grpc endpoints (instead of `registry serve --database index.db`).
+
+
+## Migration Plan
+
+The index images that exists today have the sqllite database in them, and have been built with the following Docker configuration: 
+
+```Dockerfile
+ENTRYPOINT ["/bin/opm"]
+CMD ["registry", "serve", "--database", "/database/index.db"]
+```
+
+Since the new `opm` binary will create configs on `add`, the new images will be built with the new configuration: 
+
+```Dockerfile
+ENTRYPOINT ["/bin/opm"]
+CMD ["index", "serve", "--config", "config.json"]
+```
+
+To migrate over index images built with the old configuration, a command `opm index migrate --from-index <index image> --to-index <index image>` will be introduced. This command will extrat the database from the old image, convert the data in the database into package configs, and use the configs to build an index image with the new image configuration to serve using the aggregated `config.json`. 
+The `--from-index` flag under `opm index add` is currently used to add new bundles to an existing index. This flag will be altered to check if a database exists in the index image passed as the flag's argument, and will throw an error asking the index author to fist `opm index migrate` to migrate over the old index to the new index. If the database does not exists, the flag will output a message that contains instructions on getting an index's configs, editing a config to add a new bundle, and pushing an `update` to the index, instead of carrying logic to fetch a sqllite db from an index and insert to the database (since the database will not be a part of the image anymore).
```

Comment:
we need to make sure the pipeline team knows this change is coming since they're going to have to be migrated before they pick up the new opm binary.....

Reply:
@bparees fyi updated the enhancement with an automatic migration instead of needing `opm index migrate`. This will allow for non-disruptive continuation of the use of the new opm binary for the pipeline.